data_collection_scripts/big_query_balance_data.py

```diff
@@ -72,24 +76,63 @@ def collect_data(ledgers, snapshot_dates, force_query):
+    :return: A dictionary with ledgers as keys and the corresponding start dates as values.
+    """
+    with open(hlp.ROOT_DIR / "data_collection_scripts/last_update.json") as f:
+        last_update = json.load(f)
+    from_dates = {}
+    for ledger in last_update:
+        from_dates[ledger] = hlp.increment_date(date=hlp.get_date_beginning(last_update[ledger]), by=granularity)
+    return from_dates
+
+
+def update_last_update(ledger_last_updates):
+    """
+    Update the last_update.json file with the last date for which data was collected for each ledger.
+    :param ledger_last_updates: A dictionary with the ledgers for which data was collected and the last date for which data was collected for each of them.
+    """
+    filepath = hlp.ROOT_DIR / "data_collection_scripts/last_update.json"
+    with open(filepath) as f:
+        last_update = json.load(f)
+    for ledger, date in ledger_last_updates.items():
+        if date is not None:
+            last_update[ledger] = date
+    with open(filepath, 'w') as f:
+        json.dump(last_update, f)
 
 
 if __name__ == '__main__':
     logging.basicConfig(format='[%(asctime)s] %(message)s', datefmt='%Y/%m/%d %I:%M:%S %p', level=logging.INFO)
 
     default_ledgers = hlp.get_ledgers()
-    default_snapshot_dates = hlp.get_snapshot_dates()
 
     parser = argparse.ArgumentParser()
     parser.add_argument('--ledgers', nargs="*", type=str.lower, default=default_ledgers,
                         choices=[ledger for ledger in default_ledgers], help='The ledgers to collect data for.')
-    parser.add_argument('--snapshot_dates', nargs="*", type=hlp.valid_date, default=default_snapshot_dates,
-                        help='The dates to collect data for.')
+    parser.add_argument('--to_date', type=hlp.valid_date,
+                        default=datetime.today().strftime('%Y-%m-%d'),
+                        help='The date until which to get data for (YYYY-MM-DD format). Defaults to today.')
     parser.add_argument('--force-query', action='store_true',
                         help='Flag to specify whether to query for project data regardless if the relevant data '
                              'already exist.')
     args = parser.parse_args()
 
-    snapshot_dates = [hlp.get_date_string_from_date(hlp.get_date_beginning(date)) for date in args.snapshot_dates]
-
-    collect_data(ledgers=args.ledgers, snapshot_dates=snapshot_dates, force_query=args.force_query)
+    to_date = hlp.get_date_beginning(args.to_date)
+    ledgers = args.ledgers
+    granularity = hlp.get_granularity()
+    if granularity is None:
+        ledger_snapshot_dates = {ledger: [hlp.get_date_string_from_date(to_date)] for ledger in ledgers}
+    else:
+        ledger_from_dates = get_from_dates(granularity=granularity)
+        ledger_snapshot_dates = {ledger: hlp.get_dates_between(ledger_from_dates[ledger], to_date, granularity) for ledger in ledgers}

```

Comment:
This will also throw an error if the ledger is not in the file (if that's intended, fine).